conbench/metrics.py

```diff
@@ -0,0 +1,121 @@
+https://github.com/rycus86/prometheus_flask_exporter/issues/147
+"""
+
+import logging
+import os
+
+import prometheus_client
+from prometheus_flask_exporter.multiprocess import GunicornInternalPrometheusMetrics
+
+log = logging.getLogger(__name__)
+
+
+COUNTER_GITHUB_HTTP_API_REQUESTS = prometheus_client.Counter(
+    "conbench_github_httpapi_requests_total",
+    "The total number of HTTP requests (attempted) to make to the GitHub HTTP API",
+)
+
+
+COUNTER_GITHUB_HTTP_API_REQUEST_FAILURES = prometheus_client.Counter(
+    "conbench_github_httpapi_requests_failed_total",
+    "The total number of HTTP requests to the GitHub HTTP API that failed "
+    "eventually (either immediately or after retrying).",
+)
+
+
+COUNTER_GITHUB_HTTP_API_RETRYABLE_ERRORS = prometheus_client.Counter(
+    "conbench_github_httpapi_retryable_errors_total",
+    "The total number of retryable errors observed while interacting with the "
+    "GitHub HTTP API (regardless of whether a retry was performed or not).",
+)
+
+
+COUNTER_GITHUB_HTTP_API_403RESPONSES = prometheus_client.Counter(
+    "conbench_github_httpapi_403responses_total",
+    "The total number of HTTP requests to the GitHub HTTP API that were "
+    "responded to with a 403 response. Often rate-limiting/quota related.",
+)
+
+
+GAUGE_GITHUB_HTTP_API_QUOTA_REMAINING = prometheus_client.Gauge(
+    "conbench_github_httpapi_quota_remaining",
+    "A gauge that shows the last-observed x-ratelimit-remaining response "
+    "header value.",
+    # multiprocess mode gauges are tricky!
+    # https://github.com/prometheus/client_python#multiprocess-mode-eg-gunicorn
+    # Return a single timeseries that is the minimum of the values of all
+    # processes (alive or dead).
+    multiprocess_mode="min",
+)
+
+
+# The topic of Gauge initiatlization in the Prometheus ecosystem is confusing.
+# The spec says "Gauges MUST start at 0"
+# (https://prometheus.io/docs/instrumenting/writing_clientlibs/). There are
+# client libs that thought about NaN:
+# https://github.com/micrometer-metrics/micrometer/issues/1343. With the very
+# specific setup we have (Python's client lib in multiprocessing mode) it seems
+# to report (from the start) the value 0. The value 0 however in this context
+# has a special meaning: quota exhausted. So, once this gauge reports as 0 we
+# want to know that this value came from an HTTP response. For now, set a
+# special value -1 that by convention in this case here shall mean 'not
+# initialized'. That is, if this ever shows a positive value or 0 then the
+# value was communciated within an HTTP Response.
+GAUGE_GITHUB_HTTP_API_QUOTA_REMAINING.set(-1)
+
+
+def decorate_flask_app_with_metrics(app) -> None:
+    """
+    Add flask-prometheus-exporter magic to `app`.
+
+    This mutates `app` in-place.
+    """
+    # Use `GunicornPrometheusMetrics` when spawning a separate HTTP server for
+    # the metrics scrape endpoing. Note that this sets the global singleton.
```

Comment:
nit: there's no longer a global singleton

Reply:
Thank you for paying attention!

Will update this in the next patch.

Reply:
done here: https://github.com/conbench/conbench/pull/725